x-pack/plugins/security_solution_serverless/server/endpoint/services/metering_service.ts

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { AggregationsAggregate, SearchResponse } from '@elastic/elasticsearch/lib/api/types';
+import type { ElasticsearchClient } from '@kbn/core/server';
+// import { ENDPOINT_HEARTBEAT_INDEX } from '@kbn/security-solution-plugin/common/endpoint/constants';
+import type { EndpointHeartbeat } from '@kbn/security-solution-plugin/common/endpoint/types';
+
+import type { MeteringCallbackInput } from '../../task_manager/usage_reporting_task';
+import type { UsageRecord } from '../../types';
+
+// 1 hour
+const SAMPLE_PERIOD_SECONDS = 3600;
+// const THRESHOLD_MINUTES = 30;
+
+export class EndpointMeteringService {
+  public async getUsageRecords({
+    taskId,
+    cloudSetup,
+    esClient,
+    abortController,
+    lastSuccessfulReport,
+  }: MeteringCallbackInput): Promise<UsageRecord[]> {
+    const heartbeatsResponse = await this.getHeartbeatsSince(
+      esClient,
+      abortController,
+      lastSuccessfulReport
+    );
+
+    return heartbeatsResponse.hits.hits.reduce((acc, { _source }) => {
```

Comment:
Could we move the logic inside this `reduce` in a separated function? So it is easier to read 🙂